scripts/parser.py

```diff
@@ -0,0 +1,433 @@
+        for seq_id, seq_update in sequences_to_update.items():
+            seqapi.update_sequence(
+                project=self.sample_metadata_project,
+                sequence_id=seq_id,
+                sequence_update_model=seq_update,
+            )
+
+        return ext_sample_to_internal_id
+
+    def parse_file(
+        self, reads: List[str]
+    ) -> Tuple[Union[List[List[Dict]], List[Dict]], str]:
+        """
+        Returns a tuple of:
+        1. single / list-of CWL file object(s), based on the extensions of the reads
+        2. parsed type (fastq, cram, bam)
+        """
+        if all(any(r.lower().endswith(ext) for ext in FASTQ_EXTENSIONS) for r in reads):
+            structured_fastqs = self.parse_fastqs_structure(reads)
+            files = []
+            for fastq_group in structured_fastqs:
+                files.append([self.create_file_object(f) for f in fastq_group])
+
+            return files, 'fastq'
+
+        if all(any(r.lower().endswith(ext) for ext in CRAM_EXTENSIONS) for r in reads):
+            sec_format = ['.crai', '^.crai']
+            files = []
+            for r in reads:
+                secondaries = self.create_secondary_file_objects_by_potential_pattern(
+                    r, sec_format
+                )
+                files.append(self.create_file_object(r, secondary_files=secondaries))
+
+            return files, 'cram'
+
+        if all(any(r.lower().endswith(ext) for ext in BAM_EXTENSIONS) for r in reads):
+            sec_format = ['.bai', '^.bai']
+            files = []
+            for r in reads:
+                secondaries = self.create_secondary_file_objects_by_potential_pattern(
+                    r, sec_format
+                )
+                files.append(self.create_file_object(r, secondary_files=secondaries))
+
+            return files, 'bam'
+
+        if all(any(r.lower().endswith(ext) for ext in VCFGZ_EXTENSIONS) for r in reads):
+            sec_format = ['.tbi']
+            files = []
+            for r in reads:
+                secondaries = self.create_secondary_file_objects_by_potential_pattern(
+                    r, sec_format
+                )
+                files.append(self.create_file_object(r, secondary_files=secondaries))
+
+            return files, 'gvcf'
+
+        extensions = set(os.path.splitext(r)[-1] for r in reads)
+        joined_reads = ''.join(f'\n\t{i}: {r}' for i, r in enumerate(reads))
+        raise ValueError(
+            f'Mixed, or unrecognised extensions ({", ".join(extensions)}) for reads: {joined_reads}'
+        )
+
+    @staticmethod
+    def parse_fastqs_structure(fastqs) -> List[List[str]]:
+        """
+        Takes a list of fastqs, and a set of nested lists of each R1 + R2 read.
+
+        >>> GenericParser.parse_fastqs_structure(['20210727_PROJECT1_L002_R2.fastq.gz', '20210727_PROJECT1_L002_R1.fastq.gz', '20210727_PROJECT1_L001_R2.fastq.gz', '20210727_PROJECT1_L001_R1.fastq.gz'])
```

Comment:
Note that fastq names can also have a "segment ID" in the end (e.g. `SampleName_S1_L001_R1_001.fastq.gz`):
https://support.illumina.com/help/BaseSpace_OLH_009008/Content/Source/Informatics/BS/NamingConvention_FASTQ-files-swBS.htm
Which is though always `_001`, so shouldn't cause issues.

Reply:
Ooh that's a great resource, thanks for sharing!